File: src/main/java/io/github/aalsanie/codes/ValidationResult.java
package io.github.aalsanie.codes;

import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public abstract sealed class ValidationResult permits ValidationResult.Valid, ValidationResult.Invalid {
    private ValidationResult() {
    }

    public static ValidationResult valid() {
        return Valid.INSTANCE;
    }

    public static ValidationResult invalid(Issue issue) {
        return new Invalid(List.of(Objects.requireNonNull(issue, "issue")));
    }

    public static ValidationResult invalid(List<Issue> issues) {
        return new Invalid(issues);
    }

    public static ValidationResult combine(ValidationResult... results) {
        Objects.requireNonNull(results, "results");
        return combine(List.of(results));
    }

    public static ValidationResult combine(List<ValidationResult> results) {
        Objects.requireNonNull(results, "results");
        List<Issue> issues = new ArrayList<>();
        for (ValidationResult result : results) {
            Objects.requireNonNull(result, "result");
            issues.addAll(result.issues());
        }
        return issues.isEmpty() ? Valid.INSTANCE : new Invalid(issues);
    }

    public final boolean isValid() {
        return this == Valid.INSTANCE;
    }

    public final boolean isInvalid() {
        return !isValid();
    }

    public abstract List<Issue> issues();

    public final Outcome toOutcome(OutcomeDefinition failureDefinition) {
        return toOutcome(failureDefinition, null);
    }

    public final Outcome toOutcome(
        OutcomeDefinition failureDefinition,
        @Nullable String detail
    ) {
        Objects.requireNonNull(failureDefinition, "failureDefinition");
        if (failureDefinition.getState() != OutcomeState.FAILED) {
            throw new IllegalArgumentException("validation failure definition must have FAILED state");
        }
        if (isValid()) {
            return Outcome.of(StandardOutcomes.OK, detail);
        }
        return Outcome.of(failureDefinition, detail, issues());
    }

    public static final class Valid extends ValidationResult {
        private static final Valid INSTANCE = new Valid();

        private Valid() {
        }

        @Override
        public List<Issue> issues() {
            return List.of();
        }

        @Override
        public boolean equals(Object other) {
            return other instanceof Valid;
        }

        @Override
        public int hashCode() {
            return 0;
        }

        @Override
        public String toString() {
            return "Valid";
        }
    }

    public static final class Invalid extends ValidationResult {
        private final List<Issue> issues;

        private Invalid(List<Issue> issues) {
            Objects.requireNonNull(issues, "issues");
            if (issues.isEmpty()) {
                throw new IllegalArgumentException("invalid validation result requires at least one issue");
            }
            this.issues = List.copyOf(issues);
        }

        @Override
        public List<Issue> issues() {
            return issues;
        }

        public List<Issue> getIssues() {
            return issues;
        }

        @Override
        public boolean equals(Object other) {
            return this == other || (other instanceof Invalid invalid && issues.equals(invalid.issues));
        }

        @Override
        public int hashCode() {
            return issues.hashCode();
        }

        @Override
        public String toString() {
            return "Invalid(issues=" + issues + ")";
        }
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/ValidationResult.kt
package io.github.aalsanie.codes

import java.util.Collections

sealed class ValidationResult private constructor() {
    data object Valid : ValidationResult()

    class Invalid private constructor(
        issues: List<Issue>,
    ) : ValidationResult() {
        val issues: List<Issue> = Collections.unmodifiableList(ArrayList(issues))

        init {
            require(this.issues.isNotEmpty()) {
                "invalid validation result requires at least one issue"
            }
        }

        override fun equals(other: Any?): Boolean =
            this === other ||
                (other is Invalid && issues == other.issues)

        override fun hashCode(): Int = issues.hashCode()

        override fun toString(): String = "Invalid(issues=$issues)"

        internal companion object {
            @JvmSynthetic
            internal fun create(issues: List<Issue>): Invalid = Invalid(issues)
        }
    }

    val isValid: Boolean
        get() = this === Valid

    val isInvalid: Boolean
        get() = this is Invalid

    fun issues(): List<Issue> =
        when (this) {
            Valid -> emptyList()
            is Invalid -> issues
        }

    @JvmOverloads
    fun toOutcome(
        failureDefinition: OutcomeDefinition,
        detail: String? = null,
    ): Outcome {
        require(failureDefinition.state == OutcomeState.FAILED) {
            "validation failure definition must have FAILED state"
        }
        return when (this) {
            Valid -> Outcome.of(StandardOutcomes.OK, detail)
            is Invalid -> Outcome.of(failureDefinition, detail, issues)
        }
    }

    companion object {
        @JvmStatic
        fun valid(): ValidationResult = Valid

        @JvmStatic
        fun invalid(issue: Issue): ValidationResult = Invalid.create(listOf(issue))

        @JvmStatic
        fun invalid(issues: List<Issue>): ValidationResult = Invalid.create(issues)

        @JvmStatic
        fun combine(vararg results: ValidationResult): ValidationResult = combine(results.asList())

        @JvmStatic
        fun combine(results: List<ValidationResult>): ValidationResult {
            val issues =
                results.flatMap { result ->
                    when (result) {
                        Valid -> emptyList()
                        is Invalid -> result.issues
                    }
                }
            return if (issues.isEmpty()) Valid else Invalid.create(issues)
        }
    }
}
